### PR TITLE
Feat: Basic Endpoints (CRUD)

### PR DESCRIPTION
**Pull request by Kunnu01**
Created endpoints for the channel, message and workspace module #18

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,7 +19,7 @@
     "test": "yarn test:client"
   },
   "devDependencies": {
-    "babel-eslint": "^10.0.1",
+    "babel-eslint": "10.0.1",
```
Comment (sourabhmodi20051) [inline package.json]: Why this change?

File: server/src/Controller/Channel/ChannelController.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  findOne,
+  findMany,
+  update,
+} from '../../Repositories/genericRepository';
+import successHandler from '../../libs/routes/successHandler';
+import { channelModel } from '../../Model';
+
+
+const getChannel = async (req, res, next) => {
+  try {
```
Comment (sourabhmodi20051) [inline server/src/Controller/Channel/ChannelController.js]: Can we abstract this try catch? as this is used everywhere. 
Following is one of the ways, we can look for other ways. This can be added in common utils which can be used in both client and server
https://nunoalexandre.com/2016/12/20/encapsulating-try-catch-statements
@nimish-gupta @rakeshtembhurne @radhey1851 - Are you aware of some standard way of doing this?

File: server/src/Controller/Channel/ChannelController.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    const { id } = req.params;
+    const result = await findOne(channelModel, id);
+    if (!result) {
+      const err = new Error('Channel does not exists');
```
Comment (sourabhmodi20051) [inline server/src/Controller/Channel/ChannelController.js]: Ideally, we should return 404 if the channel is not found. This text `Channel does not exist` can be moved to constants

File: server/src/Controller/Channel/ChannelController.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      next(err);
+    }
+    res
+      .status(200)
+      .send(successHandler('Channel is here', result));
```
Comment (sourabhmodi20051) [inline server/src/Controller/Channel/ChannelController.js]: `Text can be moved to constants`

File: server/src/Controller/Channel/ChannelController.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    next(error);
+  }
+};
+
+const getAllChannels = async (req, res, next) => {
+  try {
+    const result = await findMany(channelModel);
+    res
+      .status(200)
+      .send(successHandler('Channels are here', result));
```
Comment (sourabhmodi20051) [inline server/src/Controller/Channel/ChannelController.js]: results will be an array. Can we convert this to object and send as response like
`{"channels": [] }`

File: server/src/Controller/Channel/ChannelController.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  try {
+    const {
+      name,
+      workspace,
+      user,
+      isPrivate,
+      members,
+    } = req.body;
+    // Change this logic later accordingly
+    const membersArray = members.split(', ');
```
Comment (sourabhmodi20051) [inline server/src/Controller/Channel/ChannelController.js]: Why members are not an array in the request body itself?

File: server/src/Controller/Channel/ChannelController.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      name,
+      workspace,
+      user,
+      isPrivate,
+      members: membersArray,
+    };
+    const result = await create(channelModel, data);
+
+    res
+      .status(200)
```
Comment (sourabhmodi20051) [inline server/src/Controller/Channel/ChannelController.js]: For Create APIs response code should be ideally `201`

File: server/src/Controller/Channel/ChannelController.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+};
+
+const deleteChannel = async (req, res, next) => {
+  try {
+    const { id } = req.params;
+    const result = await deleteDoc(channelModel, id);
+
+    res
+      .status(200)
+      .send(successHandler('Channel deleted successfully', result));
```
Comment (sourabhmodi20051) [inline server/src/Controller/Channel/ChannelController.js]: I think error message here is typo

File: server/src/Controller/Message/MessageController.js [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  update,
+} from '../../Repositories/genericRepository';
+import successHandler from '../../libs/routes/successHandler';
+import { messageModel } from '../../Model';
+
+
+const getMessage = async (req, res, next) => {
+  try {
+    const { id } = req.params;
+    const result = await findOne(messageModel, id);
```
Comment (sourabhmodi20051) [inline server/src/Controller/Message/MessageController.js]: What is the use-case of this API? Messages will always go through sockets

File: server/src/Controller/Message/MessageController.js [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+      .status(200)
+      .send(successHandler('Message is here', result));
+  } catch (error) {
+    next(error);
+  }
+};
+
+const getAllMessages = async (req, res, next) => {
+  try {
+    const result = await findMany(messageModel);
```
Comment (sourabhmodi20051) [inline server/src/Controller/Message/MessageController.js]: I think we will never use the get all messages API. We can call API using timestamp though, like get all messages before the timestamp T.

File: server/src/Controller/Channel/ChannelController.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+      .status(200)
+      .send(successHandler('Channel is here', result));
+  } catch (error) {
+    next(error);
+  }
+};
+
+const getAllChannels = async (req, res, next) => {
+  try {
+    const result = await findMany(channelModel);
```
Comment (sourabhmodi20051) [inline server/src/Controller/Channel/ChannelController.js]: This API can have parameters like
1. user - user for whom all channels need to be provided ( this we can get from authentication middleware - for this PR it can be hardcoded for sample user maybe)
2. workspace - workspace for which all channels need to be provided

File: server/src/Controller/Channel/ChannelController.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+const getChannel = async (req, res, next) => {
+  try {
+    const { id } = req.params;
+    const result = await findOne(channelModel, id);
+    if (!result) {
+      const err = new Error('Channel does not exists');
+      next(err);
+    }
+    res
```
Comment (sourabhmodi20051) [inline server/src/Controller/Channel/ChannelController.js]: Another validation is required like
1. if this id belongs to the user or not, if it is public or private.

File: server/src/Controller/User/UserController.js [hunk 13]
```diff
@@ -38,8 +38,18 @@ const getAllUsers = async (req, res, next) => {
...
+      password,
+      bio,
+    } = req.body;
+    const data = {
+      name,
+      email,
+      password,
+      bio,
+    };
     const result = await create(userModel, data);
```
Comment (sourabhmodi20051) [inline server/src/Controller/User/UserController.js]: existence of user can be checked before creating

File: package.json (hunk 1 above)
Comment (Kunnu01) [reply package.json]: Somehow babel-eslint@10.0.2 was installed. React script works with 10.0.1 only. That's why.

File: server/src/Controller/Channel/ChannelController.js (hunk 3 above)
Comment (Kunnu01) [reply server/src/Controller/Channel/ChannelController.js]: Okay, I will change it.

File: server/src/Controller/Channel/ChannelController.js (hunk 4 above)
Comment (Kunnu01) [reply server/src/Controller/Channel/ChannelController.js]: Okay

File: server/src/Controller/Channel/ChannelController.js (hunk 5 above)
Comment (Kunnu01) [reply server/src/Controller/Channel/ChannelController.js]: Result is in format:
{
  Data: {
      data: [channels],
      count: num,
  },
  message: String,
}

File: server/src/Controller/Channel/ChannelController.js (hunk 7 above)
Comment (Kunnu01) [reply server/src/Controller/Channel/ChannelController.js]: Okay, I'll change it.

File: server/src/Controller/Message/MessageController.js (hunk 9 above)
Comment (Kunnu01) [reply server/src/Controller/Message/MessageController.js]: Won't you need this for UI purpose when you would have to show threads?

File: server/src/Controller/Channel/ChannelController.js (hunk 2 above)
Comment (nimish-gupta) [reply server/src/Controller/Channel/ChannelController.js]: You can wrap a function around try catch, which will return an array like [success, error] similar to callback style.

File: server/src/Controller/User/UserController.js (hunk 13 above)
Comment (Kunnu01) [reply server/src/Controller/User/UserController.js]: Different Validations are still to be added. There is a different issue for that.

File: server/src/Model/Message/MessageSchema.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import { Schema, model } from 'mongoose';
+import { MESSAGE_SCHEMA } from '../../Constants/constants';
+
+const messageSchema = new Schema({
+  from: String,
+  to: String,
+  message: String,
+  workspace: String,
+  channel: String,
+  createdAt: Date,
```
Comment (nimish-gupta) [inline server/src/Model/Message/MessageSchema.js]: you can use mongoose timestamps for this https://mongoosejs.com/docs/guide.html#timestamps

**Comment by Kunnu01** [comment]
I have made almost all the changes. There is a different issue for validations as  @radhey1851  needs the endpoints to work with sockets. Once these endpoints are merged, I or anyone else can pick the issue #20 and add the validations.
Please review @nimish-gupta @sourabhmodi20051 @rakeshtembhurne @radhey1851
